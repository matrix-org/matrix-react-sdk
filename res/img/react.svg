<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
