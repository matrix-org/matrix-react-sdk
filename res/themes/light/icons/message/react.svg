<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>