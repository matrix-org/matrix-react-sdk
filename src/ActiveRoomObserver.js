/*
Copyright 2017 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import RoomViewStore from './stores/RoomViewStore';

/**
 * Consumes changes from the RoomViewStore and notifies specific things
 * about when the active room changes. Unlike listening for RoomViewStore
 * changes, you can subscribe to only changes relevant to a particular
 * room.
 *
 * TODO: If we introduce an observer for something else, factor out
 * the adding / removing of listeners & emitting into a common class.
 */
class ActiveRoomObserver {
    constructor() {
        this._listeners = {};

        this._activeRoomId = RoomViewStore.getRoomId();
        // TODO: We could self-destruct when the last listener goes away, or at least
        // stop listening.
        this._roomStoreToken = RoomViewStore.addListener(this._onRoomViewStoreUpdate.bind(this));
    }

    addListener(roomId, listener) {
        if (!this._listeners[roomId]) this._listeners[roomId] = [];
        this._listeners[roomId].push(listener);
    }

    removeListener(roomId, listener) {
        if (this._listeners[roomId]) {
            const i = this._listeners[roomId].indexOf(listener);
            if (i > -1) {
                this._listeners[roomId].splice(i, 1);
            }
        } else {
            console.warn("Unregistering unrecognised listener (roomId=" + roomId + ")");
        }
    }

    _emit(roomId) {
        if (!this._listeners[roomId]) return;

        for (const l of this._listeners[roomId]) {
            l.call();
        }
    }

    _onRoomViewStoreUpdate() {
        // emit for the old room ID
        if (this._activeRoomId) this._emit(this._activeRoomId);

        // update our cache
        this._activeRoomId = RoomViewStore.getRoomId();

        // and emit for the new one
        if (this._activeRoomId) this._emit(this._activeRoomId);
    }
}

if (global.mx_ActiveRoomObserver === undefined) {
    global.mx_ActiveRoomObserver = new ActiveRoomObserver();
}
export default global.mx_ActiveRoomObserver;
