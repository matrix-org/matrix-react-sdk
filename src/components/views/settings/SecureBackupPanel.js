/*
Copyright 2018 New Vector Ltd
Copyright 2019, 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';

import {MatrixClientPeg} from '../../../MatrixClientPeg';
import { _t } from '../../../languageHandler';
import Modal from '../../../Modal';
import { isSecureBackupRequired } from '../../../utils/WellKnownUtils';
import Spinner from '../elements/Spinner';
import AccessibleButton from '../elements/AccessibleButton';
import QuestionDialog from '../dialogs/QuestionDialog';
import RestoreKeyBackupDialog from '../dialogs/security/RestoreKeyBackupDialog';
import { accessSecretStorage } from '../../../SecurityManager';
import {replaceableComponent} from "../../../utils/replaceableComponent";

@replaceableComponent("views.settings.SecureBackupPanel")
export default class SecureBackupPanel extends React.PureComponent {
    constructor(props) {
        super(props);

        this._unmounted = false;
        this.state = {
            loading: true,
            error: null,
            backupKeyStored: null,
            backupKeyCached: null,
            backupKeyWellFormed: null,
            secretStorageKeyInAccount: null,
            secretStorageReady: null,
            backupInfo: null,
            backupSigStatus: null,
            sessionsRemaining: 0,
        };
    }

    componentDidMount() {
        this._checkKeyBackupStatus();

        MatrixClientPeg.get().on('crypto.keyBackupStatus', this._onKeyBackupStatus);
        MatrixClientPeg.get().on(
            'crypto.keyBackupSessionsRemaining',
            this._onKeyBackupSessionsRemaining,
        );
    }

    componentWillUnmount() {
        this._unmounted = true;

        if (MatrixClientPeg.get()) {
            MatrixClientPeg.get().removeListener('crypto.keyBackupStatus', this._onKeyBackupStatus);
            MatrixClientPeg.get().removeListener(
                'crypto.keyBackupSessionsRemaining',
                this._onKeyBackupSessionsRemaining,
            );
        }
    }

    _onKeyBackupSessionsRemaining = (sessionsRemaining) => {
        this.setState({
            sessionsRemaining,
        });
    }

    _onKeyBackupStatus = () => {
        // This just loads the current backup status rather than forcing
        // a re-check otherwise we risk causing infinite loops
        this._loadBackupStatus();
    }

    async _checkKeyBackupStatus() {
        this._getUpdatedDiagnostics();
        try {
            const {backupInfo, trustInfo} = await MatrixClientPeg.get().checkKeyBackup();
            this.setState({
                loading: false,
                error: null,
                backupInfo,
                backupSigStatus: trustInfo,
            });
        } catch (e) {
            console.log("Unable to fetch check backup status", e);
            if (this._unmounted) return;
            this.setState({
                loading: false,
                error: e,
                backupInfo: null,
                backupSigStatus: null,
            });
        }
    }

    async _loadBackupStatus() {
        this.setState({ loading: true });
        this._getUpdatedDiagnostics();
        try {
            const backupInfo = await MatrixClientPeg.get().getKeyBackupVersion();
            const backupSigStatus = await MatrixClientPeg.get().isKeyBackupTrusted(backupInfo);
            if (this._unmounted) return;
            this.setState({
                loading: false,
                error: null,
                backupInfo,
                backupSigStatus,
            });
        } catch (e) {
            console.log("Unable to fetch key backup status", e);
            if (this._unmounted) return;
            this.setState({
                loading: false,
                error: e,
                backupInfo: null,
                backupSigStatus: null,
            });
        }
    }

    async _getUpdatedDiagnostics() {
        const cli = MatrixClientPeg.get();
        const secretStorage = cli.crypto._secretStorage;

        const backupKeyStored = !!(await cli.isKeyBackupKeyStored());
        const backupKeyFromCache = await cli.crypto.getSessionBackupPrivateKey();
        const backupKeyCached = !!(backupKeyFromCache);
        const backupKeyWellFormed = backupKeyFromCache instanceof Uint8Array;
        const secretStorageKeyInAccount = await secretStorage.hasKey();
        const secretStorageReady = await cli.isSecretStorageReady();

        if (this._unmounted) return;
        this.setState({
            backupKeyStored,
            backupKeyCached,
            backupKeyWellFormed,
            secretStorageKeyInAccount,
            secretStorageReady,
        });
    }

    _startNewBackup = () => {
        Modal.createTrackedDialogAsync('Key Backup', 'Key Backup',
            import('../../../async-components/views/dialogs/security/CreateKeyBackupDialog'),
            {
                onFinished: () => {
                    this._loadBackupStatus();
                },
            }, null, /* priority = */ false, /* static = */ true,
        );
    }

    _deleteBackup = () => {
        Modal.createTrackedDialog('Delete Backup', '', QuestionDialog, {
            title: _t('Delete Backup'),
            description: _t(
                "Are you sure? You will lose your encrypted messages if your " +
                "keys are not backed up properly.",
            ),
            button: _t('Delete Backup'),
            danger: true,
            onFinished: (proceed) => {
                if (!proceed) return;
                this.setState({loading: true});
                MatrixClientPeg.get().deleteKeyBackupVersion(this.state.backupInfo.version).then(() => {
                    this._loadBackupStatus();
                });
            },
        });
    }

    _restoreBackup = async () => {
        Modal.createTrackedDialog(
            'Restore Backup', '', RestoreKeyBackupDialog, null, null,
            /* priority = */ false, /* static = */ true,
        );
    }

    _resetSecretStorage = async () => {
        this.setState({ error: null });
        try {
            await accessSecretStorage(() => { }, /* forceReset = */ true);
        } catch (e) {
            console.error("Error resetting secret storage", e);
            if (this._unmounted) return;
            this.setState({ error: e });
        }
        if (this._unmounted) return;
        this._loadBackupStatus();
    }

    render() {
        const {
            loading,
            error,
            backupKeyStored,
            backupKeyCached,
            backupKeyWellFormed,
            secretStorageKeyInAccount,
            secretStorageReady,
            backupInfo,
            backupSigStatus,
            sessionsRemaining,
        } = this.state;

        let statusDescription;
        let extraDetailsTableRows;
        let extraDetails;
        const actions = [];
        if (error) {
            statusDescription = (
                <div className="error">
                    {_t("Unable to load key backup status")}
                </div>
            );
        } else if (loading) {
            statusDescription = <Spinner />;
        } else if (backupInfo) {
            let restoreButtonCaption = _t("Restore from Backup");

            if (MatrixClientPeg.get().getKeyBackupEnabled()) {
                statusDescription = <p>âœ… {_t("This session is backing up your keys. ")}</p>;
            } else {
                statusDescription = <>
                    <p>{_t(
                        "This session is <b>not backing up your keys</b>, " +
                        "but you do have an existing backup you can restore from " +
                        "and add to going forward.", {},
                        {b: sub => <b>{sub}</b>},
                    )}</p>
                    <p>{_t(
                        "Connect this session to key backup before signing out to avoid " +
                        "losing any keys that may only be on this session.",
                    )}</p>
                </>;
                restoreButtonCaption = _t("Connect this session to Key Backup");
            }

            let uploadStatus;
            if (!MatrixClientPeg.get().getKeyBackupEnabled()) {
                // No upload status to show when backup disabled.
                uploadStatus = "";
            } else if (sessionsRemaining > 0) {
                uploadStatus = <div>
                    {_t("Backing up %(sessionsRemaining)s keys...", { sessionsRemaining })} <br />
                </div>;
            } else {
                uploadStatus = <div>
                    {_t("All keys backed up")} <br />
                </div>;
            }

            let backupSigStatuses = backupSigStatus.sigs.map((sig, i) => {
                const deviceName = sig.device ? (sig.device.getDisplayName() || sig.device.deviceId) : null;
                const validity = sub =>
                    <span className={sig.valid ? 'mx_SecureBackupPanel_sigValid' : 'mx_SecureBackupPanel_sigInvalid'}>
                        {sub}
                    </span>;
                const verify = sub =>
                    <span className={sig.device && sig.deviceTrust.isVerified() ? 'mx_SecureBackupPanel_deviceVerified' : 'mx_SecureBackupPanel_deviceNotVerified'}>
                        {sub}
                    </span>;
                const device = sub => <span className="mx_SecureBackupPanel_deviceName">{deviceName}</span>;
                const fromThisDevice = (
                    sig.device &&
                    sig.device.getFingerprint() === MatrixClientPeg.get().getDeviceEd25519Key()
                );
                const fromThisUser = (
                    sig.crossSigningId &&
                    sig.deviceId === MatrixClientPeg.get().getCrossSigningId()
                );
                let sigStatus;
                if (sig.valid && fromThisUser) {
                    sigStatus = _t(
                        "Backup has a <validity>valid</validity> signature from this user",
                        {}, { validity },
                    );
                } else if (!sig.valid && fromThisUser) {
                    sigStatus = _t(
                        "Backup has a <validity>invalid</validity> signature from this user",
                        {}, { validity },
                    );
                } else if (sig.crossSigningId) {
                    sigStatus = _t(
                        "Backup has a signature from <verify>unknown</verify> user with ID %(deviceId)s",
                        { deviceId: sig.deviceId }, { verify },
                    );
                } else if (!sig.device) {
                    sigStatus = _t(
                        "Backup has a signature from <verify>unknown</verify> session with ID %(deviceId)s",
                        { deviceId: sig.deviceId }, { verify },
                    );
                } else if (sig.valid && fromThisDevice) {
                    sigStatus = _t(
                        "Backup has a <validity>valid</validity> signature from this session",
                        {}, { validity },
                    );
                } else if (!sig.valid && fromThisDevice) {
                    // it can happen...
                    sigStatus = _t(
                        "Backup has an <validity>invalid</validity> signature from this session",
                        {}, { validity },
                    );
                } else if (sig.valid && sig.deviceTrust.isVerified()) {
                    sigStatus = _t(
                        "Backup has a <validity>valid</validity> signature from " +
                        "<verify>verified</verify> session <device></device>",
                        {}, { validity, verify, device },
                    );
                } else if (sig.valid && !sig.deviceTrust.isVerified()) {
                    sigStatus = _t(
                        "Backup has a <validity>valid</validity> signature from " +
                        "<verify>unverified</verify> session <device></device>",
                        {}, { validity, verify, device },
                    );
                } else if (!sig.valid && sig.deviceTrust.isVerified()) {
                    sigStatus = _t(
                        "Backup has an <validity>invalid</validity> signature from " +
                        "<verify>verified</verify> session <device></device>",
                        {}, { validity, verify, device },
                    );
                } else if (!sig.valid && !sig.deviceTrust.isVerified()) {
                    sigStatus = _t(
                        "Backup has an <validity>invalid</validity> signature from " +
                        "<verify>unverified</verify> session <device></device>",
                        {}, { validity, verify, device },
                    );
                }

                return <div key={i}>
                    {sigStatus}
                </div>;
            });
            if (backupSigStatus.sigs.length === 0) {
                backupSigStatuses = _t("Backup is not signed by any of your sessions");
            }

            let trustedLocally;
            if (backupSigStatus.trusted_locally) {
                trustedLocally = _t("This backup is trusted because it has been restored on this session");
            }

            extraDetailsTableRows = <>
                <tr>
                    <td>{_t("Backup version:")}</td>
                    <td>{backupInfo.version}</td>
                </tr>
                <tr>
                    <td>{_t("Algorithm:")}</td>
                    <td>{backupInfo.algorithm}</td>
                </tr>
            </>;

            extraDetails = <>
                {uploadStatus}
                <div>{backupSigStatuses}</div>
                <div>{trustedLocally}</div>
            </>;

            actions.push(
                <AccessibleButton key="restore" kind="primary" onClick={this._restoreBackup}>
                    {restoreButtonCaption}
                </AccessibleButton>,
            );

            if (!isSecureBackupRequired()) {
                actions.push(
                    <AccessibleButton key="delete" kind="danger" onClick={this._deleteBackup}>
                        {_t("Delete Backup")}
                    </AccessibleButton>,
                );
            }
        } else {
            statusDescription = <>
                <p>{_t(
                    "Your keys are <b>not being backed up from this session</b>.", {},
                    {b: sub => <b>{sub}</b>},
                )}</p>
                <p>{_t("Back up your keys before signing out to avoid losing them.")}</p>
            </>;
            actions.push(
                <AccessibleButton key="setup" kind="primary" onClick={this._startNewBackup}>
                    {_t("Set up")}
                </AccessibleButton>,
            );
        }

        if (secretStorageKeyInAccount) {
            actions.push(
                <AccessibleButton key="reset" kind="danger" onClick={this._resetSecretStorage}>
                    {_t("Reset")}
                </AccessibleButton>,
            );
        }

        let backupKeyWellFormedText = "";
        if (backupKeyCached) {
            backupKeyWellFormedText = ", ";
            if (backupKeyWellFormed) {
                backupKeyWellFormedText += _t("well formed");
            } else {
                backupKeyWellFormedText += _t("unexpected type");
            }
        }

        let actionRow;
        if (actions.length) {
            actionRow = <div className="mx_SecureBackupPanel_buttonRow">
                {actions}
            </div>;
        }

        return (
            <div>
                <p>{_t(
                    "Back up your encryption keys with your account data in case you " +
                    "lose access to your sessions. Your keys will be secured with a " +
                    "unique Security Key.",
                )}</p>
                {statusDescription}
                <details>
                    <summary>{_t("Advanced")}</summary>
                    <table className="mx_SecureBackupPanel_statusList"><tbody>
                        <tr>
                            <td>{_t("Backup key stored:")}</td>
                            <td>{
                                backupKeyStored === true ? _t("in secret storage") : _t("not stored")
                            }</td>
                        </tr>
                        <tr>
                            <td>{_t("Backup key cached:")}</td>
                            <td>
                                {backupKeyCached ? _t("cached locally") : _t("not found locally")}
                                {backupKeyWellFormedText}
                            </td>
                        </tr>
                        <tr>
                            <td>{_t("Secret storage public key:")}</td>
                            <td>{secretStorageKeyInAccount ? _t("in account data") : _t("not found")}</td>
                        </tr>
                        <tr>
                            <td>{_t("Secret storage:")}</td>
                            <td>{secretStorageReady ? _t("ready") : _t("not ready")}</td>
                        </tr>
                        {extraDetailsTableRows}
                    </tbody></table>
                    {extraDetails}
                </details>
                {actionRow}
            </div>
        );
    }
}
