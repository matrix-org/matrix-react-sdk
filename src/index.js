/*
Copyright 2015, 2016 OpenMarket Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

var Skinner = require('./Skinner');

var t = require('counterpart');

t.registerTranslations('en', require('../languages/en'));
t.registerTranslations('pt-BR', require('../languages/pt-BR'));

t.setFallbackLocale('en');
t.setLocale(window.navigator.languages[0]);



module.exports.loadSkin = function(skinObject) {
    Skinner.load(skinObject);
};

module.exports.resetSkin = function() {
    Skinner.reset();
};

module.exports.getComponent = function(componentName) {
    return Skinner.getComponent(componentName);
};
