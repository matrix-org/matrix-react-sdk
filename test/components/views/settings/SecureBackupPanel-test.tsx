/*
Copyright 2023 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { fireEvent, render, screen, within } from "@testing-library/react";
import { mocked } from "jest-mock";

import { flushPromises, getMockClientWithEventEmitter, mockClientMethodsUser } from "../../../test-utils";
import SecureBackupPanel from "../../../../src/components/views/settings/SecureBackupPanel";
import { accessSecretStorage } from "../../../../src/SecurityManager";

jest.mock("../../../../src/SecurityManager", () => ({
    accessSecretStorage: jest.fn(),
}));

describe("<SecureBackupPanel />", () => {
    const userId = "@alice:server.org";
    const client = getMockClientWithEventEmitter({
        ...mockClientMethodsUser(userId),
        checkKeyBackup: jest.fn(),
        isKeyBackupKeyStored: jest.fn(),
        getKeyBackupEnabled: jest.fn(),
        getKeyBackupVersion: jest.fn().mockReturnValue("1"),
        isKeyBackupTrusted: jest.fn().mockResolvedValue(true),
        getClientWellKnown: jest.fn(),
        deleteKeyBackupVersion: jest.fn(),
        secretStorage: { hasKey: jest.fn() },
    });
    // @ts-ignore allow it
    client.crypto = {
        getSessionBackupPrivateKey: jest.fn(),
        isSecretStorageReady: jest.fn(),
    } as unknown as Crypto;

    const getComponent = () => render(<SecureBackupPanel />);

    beforeEach(() => {
        client.checkKeyBackup.mockResolvedValue({
            backupInfo: {
                version: "1",
                algorithm: "test",
                auth_data: {
                    public_key: "1234",
                },
            },
            trustInfo: {
                usable: false,
                sigs: [],
            },
        });

        mocked(client.secretStorage.hasKey).mockClear().mockResolvedValue(false);
        client.deleteKeyBackupVersion.mockClear().mockResolvedValue();
        client.getKeyBackupVersion.mockClear();
        client.isKeyBackupTrusted.mockClear();

        mocked(accessSecretStorage).mockClear().mockResolvedValue();
    });

    it("displays a loader while checking keybackup", async () => {
        getComponent();
        expect(screen.getByRole("progressbar")).toBeInTheDocument();
        await flushPromises();
        expect(screen.queryByRole("progressbar")).not.toBeInTheDocument();
    });

    it("handles null backup info", async () => {
        // checkKeyBackup can fail and return null for various reasons
        client.checkKeyBackup.mockResolvedValue(null);
        getComponent();
        // flush checkKeyBackup promise
        await flushPromises();

        // no backup info
        expect(screen.getByText("Back up your keys before signing out to avoid losing them.")).toBeInTheDocument();
    });

    it("suggests connecting session to key backup when backup exists", async () => {
        const { container } = getComponent();
        // flush checkKeyBackup promise
        await flushPromises();

        expect(container).toMatchSnapshot();
    });

    it("displays when session is connected to key backup", async () => {
        client.getKeyBackupEnabled.mockReturnValue(true);
        getComponent();
        // flush checkKeyBackup promise
        await flushPromises();

        expect(screen.getByText("âœ… This session is backing up your keys.")).toBeInTheDocument();
    });

    it("asks for confirmation before deleting a backup", async () => {
        getComponent();
        // flush checkKeyBackup promise
        await flushPromises();

        fireEvent.click(screen.getByText("Delete Backup"));

        const dialog = await screen.findByRole("dialog");

        expect(
            within(dialog).getByText(
                "Are you sure? You will lose your encrypted messages if your keys are not backed up properly.",
            ),
        ).toBeInTheDocument();

        fireEvent.click(within(dialog).getByText("Cancel"));

        expect(client.deleteKeyBackupVersion).not.toHaveBeenCalled();
    });

    it("deletes backup after confirmation", async () => {
        client.checkKeyBackup
            .mockResolvedValueOnce({
                backupInfo: {
                    version: "1",
                    algorithm: "test",
                    auth_data: {
                        public_key: "1234",
                    },
                },
                trustInfo: {
                    usable: false,
                    sigs: [],
                },
            })
            .mockResolvedValue(null);
        getComponent();
        // flush checkKeyBackup promise
        await flushPromises();

        fireEvent.click(screen.getByText("Delete Backup"));

        const dialog = await screen.findByRole("dialog");

        expect(
            within(dialog).getByText(
                "Are you sure? You will lose your encrypted messages if your keys are not backed up properly.",
            ),
        ).toBeInTheDocument();

        fireEvent.click(within(dialog).getByTestId("dialog-primary-button"));

        expect(client.deleteKeyBackupVersion).toHaveBeenCalledWith("1");

        // delete request
        await flushPromises();
        // refresh backup info
        await flushPromises();
    });

    it("resets secret storage", async () => {
        mocked(client.secretStorage.hasKey).mockClear().mockResolvedValue(true);
        getComponent();
        // flush checkKeyBackup promise
        await flushPromises();

        client.getKeyBackupVersion.mockClear();
        client.isKeyBackupTrusted.mockClear();

        fireEvent.click(screen.getByText("Reset"));

        // enter loading state
        expect(accessSecretStorage).toHaveBeenCalled();
        await flushPromises();

        // backup status refreshed
        expect(client.getKeyBackupVersion).toHaveBeenCalled();
        expect(client.isKeyBackupTrusted).toHaveBeenCalled();
    });
});
