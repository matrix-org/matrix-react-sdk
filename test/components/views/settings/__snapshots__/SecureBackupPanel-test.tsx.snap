// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SecureBackupPanel /> suggests connecting session to key backup when backup exists 1`] = `
<div>
  <div
    class="mx_SettingsSubsection_text"
  >
    Back up your encryption keys with your account data in case you lose access to your sessions. Your keys will be secured with a unique Security Key.
  </div>
  <div
    class="mx_SettingsSubsection_text"
  >
    <span>
      This session is 
      <b>
        not backing up your keys
      </b>
      , but you do have an existing backup you can restore from and add to going forward.
    </span>
  </div>
  <div
    class="mx_SettingsSubsection_text"
  >
    Connect this session to key backup before signing out to avoid losing any keys that may only be on this session.
  </div>
  <details>
    <summary>
      Advanced
    </summary>
    <table
      class="mx_SecureBackupPanel_statusList"
    >
      <tr>
        <th
          scope="row"
        >
          Backup key stored:
        </th>
        <td>
          not stored
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          Backup key cached:
        </th>
        <td>
          not found locally
          
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          Secret storage public key:
        </th>
        <td>
          not found
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          Secret storage:
        </th>
        <td>
          not ready
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          Backup version:
        </th>
        <td>
          1
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          Algorithm:
        </th>
        <td>
          test
        </td>
      </tr>
    </table>
    
    <div>
      Backup is not signed by any of your sessions
    </div>
    <div />
  </details>
  <div
    class="mx_SecureBackupPanel_buttonRow"
  >
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_primary"
      role="button"
      tabindex="0"
    >
      Connect this session to Key Backup
    </div>
    <div
      class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_danger"
      role="button"
      tabindex="0"
    >
      Delete Backup
    </div>
  </div>
</div>
`;
